linux/src/threadpool_linux.c

```diff
@@ -0,0 +1,303 @@
+
+        srw_lock_destroy(threadpool->srw_lock);
+        free(threadpool);
+    }
+}
+
+int threadpool_schedule_work(THREADPOOL_HANDLE threadpool, THREADPOOL_WORK_FUNCTION work_function, void* work_function_context)
+{
+    int result;
+    if (threadpool == NULL || work_function == NULL)
+    {
+        LogError("Invalid arguments: THREADPOOL_HANDLE threadpool: %p, THREADPOOL_TASK_FUNC work_function: %p", threadpool, work_function);
+        result = MU_FAILURE;
+    }
+    else
+    {
+        THREADPOOL_TASK* task_item = malloc(sizeof(THREADPOOL_TASK));
+        if (task_item == NULL)
+        {
+            LogError("Failure allocating THREADPOOL_TASK structure");
+            result = MU_FAILURE;
+        }
+        else
+        {
+            task_item->next = NULL;
+            task_item->task_param = work_function_context;
+            task_item->task_func = work_function;
+
+            // Need to add a lock here
+            srw_lock_acquire_exclusive(threadpool->srw_lock);
+
+            if (threadpool->task_list == NULL)
+            {
+                threadpool->task_list = task_item;
+            }
+            else
+            {
+                threadpool->task_list_last->next = task_item;
+            }
+            threadpool->task_list_last = task_item;
+            (void)interlocked_increment(&threadpool->task_count);
+
+            srw_lock_release_exclusive(threadpool->srw_lock);
+
+            // unlock
+
+            if (interlocked_add(&threadpool->pool_state, 0) == POOL_STATE_IDLE)
+            {
+                // Idle signal that we have work
+                wake_by_address_single(&threadpool->task_count);
+                result = 0;
+                goto all_ok;
+            }
+            else
+            {
+                // Check to see
+                if (threadpool->thread_count < threadpool->max_thread_count)
+                {
+                    if (ThreadAPI_Create(&threadpool->thread_handle_list[threadpool->list_index], threadpool_work_func, threadpool) != THREADAPI_OK)
+                    {
+                        LogError("Failure creating thread");
+                    }
+                    else
+                    {
+                        (void)interlocked_increment(&threadpool->thread_count);
+                        threadpool->list_index++;
+                        result = 0;
+                        goto all_ok;
+                    }
+                }
+                else
+                {
+                    result = 0;
+                    goto all_ok;
+                }
+            }
+            free(task_item);
+        }
+        result = MU_FAILURE;
+    }
+all_ok:
+    return result;
+}
+
+int threadpool_open_async(THREADPOOL_HANDLE threadpool, ON_THREADPOOL_OPEN_COMPLETE on_open_complete, void* on_open_complete_context)
+{
+    (void)threadpool;
+    (void)on_open_complete;
+    (void)on_open_complete_context;
+    return MU_FAILURE;
```

Comment:
No open at all?
Seems like we need to fill this in otherwise this won't work with zrpc.